### ISSUE TITLE
http/2 arbitrary hangs in Windows since connection filter refactor

### ISSUE THREAD

**Issue by jay** [curl/curl#10253]
### I did this

@icing I am seeing arbitrary hangs during http/2 transfers in Windows and I've bisected it to 71b7e016 `lib: connect/h2/h3 refactor`. I think it has something to do with the pre-receive behavior that is on by default in Windows.

When I use curl on Windows built with `-DDONT_USE_RECV_BEFORE_SEND_WORKAROUND` I cannot reproduce. When I use curl on Ubuntu built with `-DUSE_RECV_BEFORE_SEND_WORKAROUND` I can reproduce. The hangs seem arbitrary but I've noticed smaller transfers seem more likely to hang. For example https://httpbin.org/bytes/1 (get 1 random byte) it takes about half a dozen attempts for me to see a hang. I have a local nginx server where the odds of curl hanging are better if I make up a non-existent path like https://localhost/asdf. It looks like timing has something to do with it as well.

<details>
  <summary>Click to expand sample output from a hang in Ubuntu</summary>
  
~~~
> GET /bytes/1 HTTP/2
> Host: httpbin.org
> user-agent: curl/7.87.1-DEV
> accept: */*
> 
* STATE: DO => DID handle 0x622000019908; line 2198 (connection #0)
* multi changed, check CONNECT_PEND queue
* STATE: DID => PERFORMING handle 0x622000019908; line 2317 (connection #0)
* http2_recv: easy 0x622000019908 (stream 1) win 33554432/65535
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
* nread=40
* Got SETTINGS
* MAX_CONCURRENT_STREAMS == 128
* ENABLE_PUSH == TRUE
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
* h2_process_pending_input: All data in connection buffer processed
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS header, Supplemental data (23):
* http2_recv returns AGAIN for stream 1
* multi changed, check CONNECT_PEND queue
~~~
</details>

The pre-receive behavior is basically the socket is read, via [pre_receive_plain](https://github.com/curl/curl/blob/d7b56ac75cd6e2ec6d501780e46f8bb3046ebd9f/lib/sendf.c#L149), before a plain send, because in Windows when a send fails the Windows sockets library (winsock) wipes out any data in the winsock receive buffer for that socket. That is, libcurl is making an attempt to read received data held by the OS in case the send fails, to increase the chances of delivering a server error message to the user.

The pre-receive buffer is checked by transfer's data_pending via cfilters' Curl_conn_data_pending:

https://github.com/curl/curl/blob/d7b56ac75cd6e2ec6d501780e46f8bb3046ebd9f/lib/cfilters.c#L449-L467

I put a breakpoint on Curl_conn_data_pending but it's never called. What's more interesting though is if I go to a good commit it's never called either. So then I started looking into pre-receive and how it functions, to confirm it is the problem.

To decide whether or not to actually pre-receive data the pre_receive_plain checks to make sure the transfer is HTTP. I assume this necessary, for example one reason I can think of is outside of the library say you have a user using [CURLOPT_CONNECT_ONLY](https://curl.se/libcurl/c/CURLOPT_CONNECT_ONLY.html) with curl_easy_send/curl_easy_recv, and waiting with select, they couldn't determine whether data has been pre-received or not.

Also pre_receive_plain checks the recv handler function pointer to make sure it's Curl_conn_recv. I assume it wanted to only cache raw recv and I don't know why that is necessary.

https://github.com/curl/curl/blob/d7b56ac75cd6e2ec6d501780e46f8bb3046ebd9f/lib/sendf.c#L157-L164

Before the refactor, if HTTP/2 is selected by the server then conn->recv[num] is http2_recv, and since http2_recv != Curl_conn_recv then no pre-receive takes place.

After the refactor, if HTTP/2 is selected by the server then conn->recv[num] remains Curl_conn_recv during the HTTP/2 session and a pre-receive can take place at any time.

### I expected the following

I don't know how this is supposed to be managed under connection filters. This pre-receive behavior has caused some issues in the past, and I have to wonder if it's worth maintaining because bugs like this are hard to find and don't show up in tests.

### curl/libcurl version

master d7b56ac7 2023-01-06 on Ubuntu with CPPFLAGS=-DUSE_RECV_BEFORE_SEND_WORKAROUND
~~~
curl 7.87.1-DEV (x86_64-pc-linux-gnu) libcurl/7.87.1-DEV OpenSSL/3.0.7 zlib/1.2.8 nghttp2/1.50.0 librtmp/2.3
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL threadsafe TLS-SRP TrackMemory UnixSockets
~~~

master d7b56ac7 2023-01-06 on Windows (Windows builds define USE_RECV_BEFORE_SEND_WORKAROUND by default)
~~~
curl 7.87.1-DEV (i386-pc-win32) libcurl/7.87.1-DEV OpenSSL/1.1.1s nghttp2/1.50.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile NTLM SSL Unicode UnixSockets
~~~

**Comment by icing** [comment]
@jay, thanks for deep diving into the mysteries of pre-receive. I agree that this is a painful workaround and makes one wonder if it is worth it. Since, I do not run Windows, I cannot judge. I supposed it was added for good reasons.

In the current architecture, we should move the `post_poned` data into the cfilter handling the socket (candidate: `struct cf_socket_ctx`). This has the following benefits:

* it is the lowest level anyone should read from (we still have code that uses `conn->sock[index]` directly, but will migrate those parts).
* it is cleared/freed when the cfilter is closed, e.g. the socket is closed/destroyed.
* it is the correct place if we ever do consider some "buffering" on top of the socket. Especially on the receive direction, this could make sense. TLS backends often make very small reads.

**Comment by icing** [comment]
See #10280 for a workaround and an analysis of the problem in the PR description.

**Comment by icing** [comment]
@jay could you check if #10280 works for you? I added the changes mentioned in my previous comment.

**Comment by jay** [comment]
AFAICT this issue was solved by 5651a36d (#10280).

I've proposed #10409 to disable by default the recv-before-send behavior because it can lead to bugs like this.
